www/modloader/early_loader.js

```diff
@@ -87,6 +90,178 @@
         return
     }
 
+    /*
+     * Abstract class for parsing mod file data sources BEFORE they are handed off to the VFS.
+     * These Mod File Parsers are responsible for providing a Buffer that the Mod File then hands off to the VFS in a dataSource.
+     */
+    class ModFileParser {
+        constructor(modFile) {
+            this.modFile = modFile;
+        }
+
+        async parse() {
+            throw new Error('Unimplemented');
+        }
+    }
+
+    /*
+     * Mod File Parser for ES Module plugins. Parses ES Modules with rollup and returns a Buffer.
+     */
+    class ESModuleParser extends ModFileParser {
+        async parse() {
+            const rollupInstance = await $modLoader.$rollup({
+                input: this.entryPointRollupFileId,
+                plugins: [this.rollupPlugin],
+            });
+
+            const result = await rollupInstance.generate({});
+            const generatedCode = result.output[0].code;
+
+            return ESModuleParser.bufferFromGeneratedCode(generatedCode);
+        }
+
+        get entryPointRollupFileId() {
+            return ESModuleParser.buildRollupFileId("mod", this.modFile.srcFile);
+        }
+
+        get rollupPlugin() {
+            return {
+                name: "OneloaderResolver",
+                resolveId: (source, importer) => {
+                    if (importer) {
+                        const { fingerprint, filename } = ESModuleParser.parseRollupFileId(importer);
+                        if (fingerprint && filename) return ESModuleParser.buildRollupFileId(fingerprint, path.relative(path.dirname(filename), source));
+                    } else if (ESModuleParser.parseRollupFileId(source).filename) {
+                        return source;
+                    }
+                    return null;
+                },
+                load: async id => {
+                    const { fingerprint, filename } = ESModuleParser.parseRollupFileId(id);
+                    if (fingerprint !== "mod") return;
+
+                    const buffer = await this.modFile.mod.readFile(path.join(this.modFile.mod.rootPath, this.modFile.srcPath, filename));
+                    return buffer.toString("utf-8");
+                },
+            };
+        }
+
+        static bufferFromGeneratedCode(generatedCode) {
+            const wrappedCode = `(function(){${generatedCode}})()`;
+            return Buffer.from(wrappedCode);
```

Comment:
It's probably a good idea to specify encoding in Buffer.from. This would have to be a codebase-wide change, just a note for now.